pkg/workload/workload.go

```diff
@@ -520,3 +523,37 @@ func RemoveFinalizer(ctx context.Context, c client.Client, wl *kueue.Workload) e
 	}
 	return nil
 }
+
+// AdmissionChecksForWorkload return AdmissionChecks that should be assigned to a specific Workload based on
+// ClusterQueue configuration and ResourceFlavors
+func AdmissionChecksForWorkload(log *logr.Logger, wl *kueue.Workload, cq *kueue.ClusterQueue) []string {
+	if len(cq.Spec.AdmissionChecks) != 0 {
+		return cq.Spec.AdmissionChecks
+	}
+
+	// Kueue sets AdmissionChecks first based on ClusterQueue configuration and at this point Workload has no
+	// ResourceFlavors assigned, so we cannot match AdmissionChecks to ResourceFlavor.
+	// After Quota is reserved, another reconciliation happens and we can match AdmissionChecks to ResourceFlavors
+	if wl.Status.Admission == nil {
+		log.V(2).Info("Workload has no Admission", "Workload", wl)
+		return nil
+	}
+
+	wlFlavors := make([]kueue.ResourceFlavorReference, 0)
+	for _, podSet := range wl.Status.Admission.PodSetAssignments {
+		wlFlavors = append(wlFlavors, expmaps.Values(podSet.Flavors)...)
+	}
+	// Since different resources may have the same ResourceFlavor, we need to remove duplicates
+	slices.Sort(wlFlavors)
+	wlFlavors = slices.Compact(wlFlavors)
+
+	admissionCheckNames := make([]string, 0)
+	for _, acStrategy := range cq.Spec.AdmissionCheckStrategy {
+		for _, flavor := range wlFlavors {
+			if len(acStrategy.OnFlavors) == 0 || slices.Contains(acStrategy.OnFlavors, string(flavor)) {
+				admissionCheckNames = append(admissionCheckNames, acStrategy.Name)
+			}
+		}
+	}
```

Comment:
Should we reduce the number of loops? The current implementation has 7 loop points:

1. `for _, podSet := range wl.Status.Admission.PodSetAssignments {`
2. `expmaps.Values(podSet.Flavors)`
3. `slices.Sort(wlFlavors)`
4. `slices.Compact(wlFlavors)`
5. `for _, acStrategy := range cq.Spec.AdmissionCheckStrategy {`
6. `for _, flavor := range wlFlavors {`
7. `slices.Contains(acStrategy.OnFlavors, string(flavor))`

I guess that using `sets.Set` allows us to reduce loops like this.
Also, I guess that replacing the returned type with `sets.Set` would allow us to avoid creating them in the scheduler.

```golang
func AdmissionChecksForWorkload(log *logr.Logger, wl *kueue.Workload, cq *kueue.ClusterQueue) sets.Set[string] {
[...]

assinedFlvs := sets.New[string]()
for _, podSet := range wl.Status.Admission.PodSetAssignments {
	for _, fName := range podSet.Flavor {
		assinedFlvs.Insert(fName)
	}
}

acNames := sets.New[string]()
for _, acStrategy := range cq.Spec.AdmissionCheckStrategy {
	if len(acStrategy.OnFlavors) == 0 {
		acNames.Insert(acStrategy.Name)
		continue
	}
	for _, fName := range acStrategy.OnFlavors {
		if assignedFlvs.Has(fName) {
			acNames.Insert(acStrategy.Name)
		}
	}
}

return acNames
```

Note that the implementation above is just a reference. So, I think that we need to refactor it.

Reply:
Thanks for catching that. Fixed